alg/teca_gradient.cxx

```diff
@@ -0,0 +1,464 @@
+    const num_t *f, unsigned long n_lon,
+    unsigned long n_lat, bool periodic_lon=true)
+{
+    size_t n_bytes = n_lat*sizeof(num_t);
+    num_t *delta_x = static_cast<num_t*>(malloc(n_bytes));
+
+    // delta lon as a function of latitude
+    num_t delta_x_equator = num_t(2)*(lon[1] - lon[0]) * deg_to_rad<num_t>() * earth_radius<num_t>();
+    for (unsigned long j = 0; j < n_lat; ++j)
+        delta_x[j] = delta_x_equator * cos(lat[j] * deg_to_rad<num_t>());
+
+    // delta lat
+    num_t delta_y = (lat[1] - lat[0]) * deg_to_rad<num_t>() * earth_radius<num_t>();
+    num_t dy = num_t(2)*delta_y;
+
+    unsigned long max_i = n_lon - 1;
+    unsigned long max_j = n_lat - 1;
+
+    // gradient
+    for (unsigned long j = 1; j < max_j; ++j)
+    {
+        unsigned long jj = j*n_lon;
+        const num_t *f_y2 = f + jj + n_lon;
+        const num_t *f_y0 = f + jj - n_lon;
+        const num_t *f_x2 = f + jj + 1;
+        const num_t *f_x0 = f + jj - 1;
+        num_t *gx = grad_x + jj;
+        num_t *gy = grad_y + jj;
+        num_t dx = delta_x[j];
+
+        for (unsigned long i = 1; i < max_i; ++i)
+        {
+            gx[i] = (f_x2[i] - f_x0[i]) / dx;
+            gy[i] = (f_y2[i] - f_y0[i]) / dy;
+        }
+    }
+
+    if (periodic_lon)
+    {
+        // periodic in longitude
+        for (unsigned long j = 1; j < max_j; ++j)
+        {
+            unsigned long jj = j*n_lon;
+            const num_t *f_y2 = f + jj + n_lon;
+            const num_t *f_y0 = f + jj - n_lon;
+            const num_t *f_x2 = f + jj + 1;
+            const num_t *f_x0 = f + jj + max_i;
+            num_t *gx = grad_x + jj;
+            num_t *gy = grad_y + jj;
+            num_t dx = delta_x[j];
+
+            gx[0] = (f_x2[0] - f_x0[0]) / dx;
+            gy[0] = (f_y2[0] - f_y0[0]) / dx;
+        }
+
+        for (unsigned long j = 1; j < max_j; ++j)
+        {
+            unsigned long jj = j*n_lon;
+            const num_t *f_y2 = f + jj + max_i + n_lon;
+            const num_t *f_y0 = f + jj + max_i - n_lon;
+            const num_t *f_x2 = f + jj;
+            const num_t *f_x0 = f + jj + max_i - 1;
+            num_t *gx = grad_x + jj + max_i;
+            num_t *gy = grad_y + jj + max_i;
+            num_t dx = delta_x[j];
+
+            gx[0] = (f_x2[0] - f_x0[0]) / dx;
+            gy[0] = (f_y2[0] - f_y0[0]) / dx;
+        }
+    }
+    else
+    {
+        // zero it out
+        for (unsigned long j = 1; j < max_j; ++j)
+        {
+            grad_x[j*n_lon] = num_t();
+            grad_y[j*n_lon] = num_t();
+        }
+
+        for (unsigned long j = 1; j < max_j; ++j)
+        {
+            grad_x[j*n_lon + max_i] = num_t();
+            grad_y[j*n_lon + max_i] = num_t();
+        }
+    }
+
+    // extend y-derivative values into lat boundaries
+    num_t *dest = grad_y;
+    num_t *src = grad_y + n_lon;
+    for (unsigned long i = 0; i < n_lon; ++i)
+        dest[i] = src[i+n_lon];
```

Comment:
Wouldn't it be dest[i] = src[i] in line 130?

Reply:
I see what you're saying.  It does look like that, but I'm confused because this code reproduced a python version of the same boundary condition (implemented totally differently, so it's unlikely the same mistake was made twice).

@burlen, this is based on `teca_vorticity.cxx` (see [line 117](https://github.com/LBL-EESA/TECA/blob/83512cb29765d1f578a2615a4a91e0262af5b446/alg/teca_vorticity.cxx#L117)), so if there's a problem here, there is a problem there too.  I always get turned around when thinking about pointer math though.  What do you think?

Reply:
I can confirm that this was a pointer math bug. (And it turns out there was a bug in the code I used to compare with the  python-generated version, such that I wasn't actually comparing TECA's gradient at all.)

I've also submitted a fix to `teca_vorticity.cxx`.

Reply:
@taobrienlbl sorry for late reply, agree. Good catch @amandasd!